### PR TITLE
[Yul] Introduce LiteralRematerializer and thus simplify StructuralSimplifier.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libyul/optimiser/Rematerialiser.h [hunk 1]
```diff
@@ -64,8 +64,30 @@ class Rematerialiser: public DataFlowAnalyzer
 	using ASTModifier::visit;
 	void visit(Expression& _e) override;
 
+	bool m_rematExactlyLiterals = false;
```
Comment (ekpyron) [inline libyul/optimiser/Rematerialiser.h]: Unused remnant of an older version I guess?

File: libyul/optimiser/StructuralSimplifier.h [hunk 2]
```diff
@@ -30,22 +30,26 @@ namespace yul
...
  *
- * Prerequisite: Disambiguator, ForLoopInitRewriter.
+ * The LiteralRematerialiser should be run before this.
+ *
+ * Prerequisite: Disambiguator.
  *
  * Important: Can only be used on EVM code.
  */
-class StructuralSimplifier: public DataFlowAnalyzer
+class StructuralSimplifier: public ASTModifier
```
Comment (ekpyron) [inline libyul/optimiser/StructuralSimplifier.h:39]: This is a really nice side effect!

File: test/libyul/yulOptimizerTests/fullSuite/aztec.yul [hunk 3]
```diff
@@ -238,8 +238,8 @@
 //         let notes := add(0x04, calldataload(0x04))
 //         let m := calldataload(0x24)
 //         let n := calldataload(notes)
-//         let gen_order := 0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001
-//         let challenge := mod(calldataload(0x44), gen_order)
+//         let _1 := 0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001
```
Comment (ekpyron) [inline test/libyul/yulOptimizerTests/fullSuite/aztec.yul:241]: Would be nice to keep the name (i.e. pass it as hint somewhere), but not too important I guess...

File: libyul/optimiser/StructuralSimplifier.h [hunk 4]
```diff
@@ -30,22 +30,26 @@ namespace yul
...
-	using DataFlowAnalyzer::operator();
+	using ASTModifier::operator();
 	void operator()(Block& _block) override;
 private:
 	void simplify(std::vector<Statement>& _statements);
 	bool expressionAlwaysTrue(Expression const& _expression);
 	bool expressionAlwaysFalse(Expression const& _expression);
 	boost::optional<dev::u256> hasLiteralValue(Expression const& _expression) const;
+
+	Dialect const& m_dialect;
```
Comment (ekpyron) [inline libyul/optimiser/StructuralSimplifier.h]: Some of the compilers claim that this can be removed...

**Review by ekpyron — COMMENTED** [review]
Also has a failing command line test.

File: test/libyul/yulOptimizerTests/fullSuite/aztec.yul (hunk 3 above)
Comment (chriseth) [reply test/libyul/yulOptimizerTests/fullSuite/aztec.yul:241]: I know, but we can't this way.

**Comment by chriseth** [comment]
Updated.

**Review by ekpyron — APPROVED** [review]
Looks good to me, if tests pass now!

**Pull request merged**